File: src/rust/common/lightsss.rs
use std::collections::VecDeque;
use std::time::{Duration, Instant};

use log::{info, warn};

const FORK_INTERVAL: Duration = Duration::from_secs(3);
const SLOT_SIZE: usize = 2;

#[derive(Debug, PartialEq)]
pub enum ForkResult {
    Ok,
    Child { end_cycles: u64 },
}

/// Every process has its own `LightSSS` instance after fork.
/// - `pid_slots`: only used by parent, stores `(child_pid, pipe_write_fd)`
/// - `last_fork`: timestamp of last fork (parent) or of creation (child)
pub struct LightSSS {
    /// (rfd, wfd)
    pipe: (i32, i32),
    pid_slots: VecDeque<i32>,
    is_child_process: bool,
    last_fork: Instant,
}

impl LightSSS {
    pub fn new() -> Self {
        let mut fds = [0i32; 2];
        let ret = unsafe { libc::pipe(fds.as_mut_ptr()) };
        assert!(ret == 0, "pipe() failed");
        Self {
            pipe: (fds[0], fds[1]),
            pid_slots: VecDeque::new(),
            is_child_process: false,
            last_fork: Instant::now(),
        }
    }

    pub fn should_fork(&self) -> bool {
        !self.is_child_process && self.last_fork.elapsed() >= FORK_INTERVAL
    }

    pub fn is_child(&self) -> bool {
        self.is_child_process
    }

    pub fn do_fork(&mut self) -> ForkResult {
        // queue full, pop one
        if self.pid_slots.len() >= SLOT_SIZE {
            let oldest_pid = self.pid_slots.pop_back().unwrap();
            unsafe { libc::kill(oldest_pid, libc::SIGKILL) };
            unsafe { libc::waitpid(oldest_pid, std::ptr::null_mut(), 0) };
        }

        let pid = unsafe { libc::fork() };
        assert!(pid >= 0, "fork() failed");

        if pid > 0 {
            self.pid_slots.push_front(pid);
            self.last_fork = Instant::now();
            info!(
                "[lightsss] forked checkpoint pid={pid}, slots={}",
                self.pid_slots.len()
            );
            ForkResult::Ok
        } else {
            self.pid_slots.clear(); // child, no use of pid_slots
            self.is_child_process = true;
            let mut buf = [0u8; size_of::<u64>()];
            let ret = unsafe {
                libc::read(
                    self.pipe.0,
                    buf.as_mut_ptr() as *mut libc::c_void,
                    size_of::<u64>(),
                )
            };
            assert!((ret as usize) == size_of::<u64>(), "read() failed");
            let end_cycles = u64::from_ne_bytes(buf);
            let my_pid = unsafe { libc::getpid() };
            info!("[lightsss] child pid={my_pid} woke up, will dump wave until cycle {end_cycles}");
            ForkResult::Child { end_cycles }
        }
    }

    pub fn wakeup_child(&mut self, cycles: u64) {
        if self.pid_slots.is_empty() {
            warn!("[lightsss] no checkpoint to wake up");
            return;
        }

        let oldest_pid = self.pid_slots.pop_back().unwrap();

        for &pid in &self.pid_slots {
            unsafe { libc::kill(pid, libc::SIGKILL) };
            unsafe { libc::waitpid(pid, std::ptr::null_mut(), 0) };
        }

        let buf: [u8; size_of::<u64>()] = cycles.to_ne_bytes();
        let n = unsafe {
            libc::write(
                self.pipe.1,
                buf.as_ptr() as *const libc::c_void,
                size_of::<u64>(),
            )
        };
        assert!((n as usize) == size_of::<u64>(), "pipe write failed");

        info!("[lightsss] waking child pid={oldest_pid}, waiting...");
        unsafe { libc::waitpid(oldest_pid, std::ptr::null_mut(), 0) };
        info!("[lightsss] child finished");

        self.pid_slots.clear();
    }

    pub fn do_clear(&mut self) {
        unsafe {
            libc::close(self.pipe.0);
        }
        unsafe {
            libc::close(self.pipe.1);
        }
        while let Some(pid) = self.pid_slots.pop_back() {
            unsafe { libc::kill(pid, libc::SIGKILL) };
            unsafe { libc::waitpid(pid, std::ptr::null_mut(), 0) };
        }
    }
}

impl Drop for LightSSS {
    fn drop(&mut self) {
        if !self.is_child_process {
            self.do_clear();
        }
    }
}

File: src/cpp/common/lightsss.cc
#include "common/lightsss.h"

#include <cassert>
#include <chrono>
#include <cstring>

#include <signal.h>
#include <sys/wait.h>
#include <unistd.h>

#include "absl/log/log.h"

namespace npc {

static constexpr uint64_t kForkIntervalMs = 3000;
static constexpr size_t kSlotSize = 2;

uint64_t LightSSS::now_ms() {
  using namespace std::chrono;
  return duration_cast<milliseconds>(
             steady_clock::now().time_since_epoch())
      .count();
}

LightSSS::LightSSS() {
  int fds[2];
  int ret = pipe(fds);
  assert(ret == 0);
  (void)ret;
  pipe_rfd_ = fds[0];
  pipe_wfd_ = fds[1];
  last_fork_ms_ = now_ms();
}

LightSSS::~LightSSS() {
  if (!is_child_process_) {
    do_clear();
  }
}

bool LightSSS::should_fork() const {
  return !is_child_process_ && (now_ms() - last_fork_ms_ >= kForkIntervalMs);
}

ForkResult LightSSS::do_fork() {
  if (pid_slots_.size() >= kSlotSize) {
    int oldest_pid = pid_slots_.back();
    pid_slots_.pop_back();
    kill(oldest_pid, SIGKILL);
    waitpid(oldest_pid, nullptr, 0);
  }

  pid_t pid = fork();
  assert(pid >= 0);

  if (pid > 0) {
    // Parent
    pid_slots_.push_front(pid);
    last_fork_ms_ = now_ms();
    LOG(INFO) << "[lightsss] forked checkpoint pid=" << pid
              << ", slots=" << pid_slots_.size();
    return ForkResult{false, 0};
  }

  // Child
  pid_slots_.clear();
  is_child_process_ = true;

  uint64_t end_cycles = 0;
  ssize_t n = read(pipe_rfd_, &end_cycles, sizeof(end_cycles));
  assert(n == sizeof(end_cycles));
  (void)n;

  pid_t my_pid = getpid();
  LOG(INFO) << "[lightsss] child pid=" << my_pid
            << " woke up, will dump wave until cycle " << end_cycles;
  return ForkResult{true, end_cycles};
}

void LightSSS::wakeup_child(uint64_t cycles) {
  if (pid_slots_.empty()) {
    LOG(WARNING) << "[lightsss] no checkpoint to wake up";
    return;
  }

  // pop back
  int oldest_pid = pid_slots_.back();
  pid_slots_.pop_back();

  for (int pid : pid_slots_) {
    kill(pid, SIGKILL);
    waitpid(pid, nullptr, 0);
  }

  ssize_t n = write(pipe_wfd_, &cycles, sizeof(cycles));
  assert(n == sizeof(cycles));
  (void)n;

  LOG(INFO) << "[lightsss] waking child pid=" << oldest_pid << ", waiting...";
  waitpid(oldest_pid, nullptr, 0);
  LOG(INFO) << "[lightsss] child finished";

  pid_slots_.clear();
}

void LightSSS::do_clear() {
  close(pipe_rfd_);
  close(pipe_wfd_);
  while (!pid_slots_.empty()) {
    int pid = pid_slots_.back();
    pid_slots_.pop_back();
    kill(pid, SIGKILL);
    waitpid(pid, nullptr, 0);
  }
}

}  // namespace npc
